components/places/src/storage/tags.rs

```diff
@@ -0,0 +1,317 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use super::{fetch_page_info, TAG_LENGTH_MAX};
+use crate::error::{InvalidPlaceInfo, Result};
+use rusqlite::Connection;
+use sql_support::ConnExt;
+use url::Url;
+
+/// Checks the validity of the specified tag. On success, the result is the
+/// string value which should be used for the tag, or failure indicates why
+/// the tag is not valid.
+pub fn validate_tag(t: &str) -> Result<&str> {
+    // Drop empty and oversized tags.
+    let t = t.trim();
+    if t.len() == 0 || t.len() > TAG_LENGTH_MAX || t.find(|c: char| c.is_whitespace()).is_some() {
+        Err(InvalidPlaceInfo::InvalidTag(t.to_string()).into())
+    } else {
+        Ok(t)
+    }
+}
+
+/// Tags the specified URL.
+///
+/// # Arguments
+///
+/// * `conn` - A database connection on which to operate.
+///
+/// * `url` - The URL to tag.
+///
+/// * `tag` - The tag to add for the URL.
+///
+/// # Returns
+///
+/// There is no success return value.
+pub fn tag_url(conn: &Connection, url: &Url, tag: String) -> Result<()> {
+    let tag = validate_tag(&tag)?;
+    let tx = conn.unchecked_transaction()?;
+
+    // This function will not create a new place.
+    // Fetch the place id, so we (a) avoid creating a new tag when we aren't
+    // going to reference it and (b) to avoid a sub-query.
+    let place_id = match fetch_page_info(conn, url)? {
+        Some(info) => info.page.row_id,
+        None => return Err(InvalidPlaceInfo::NoItem(url.to_string()).into()),
```

Comment:
Usually NoItem is used for Guids. In general, I think we should keep PII like specific URLs out of error information, since we might want to send it in telemetry.